File: CS/AsynchronousFindInTreeList/ExpandingToSpecificNode.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Data;
using DevExpress.XtraTreeList.Nodes;
using DevExpress.XtraTreeList;
using System.ComponentModel;
using System.Threading;
using System.Windows.Forms;
using DevExpress.XtraTreeList.Nodes.Operations;

namespace AsynchronousFindInTreeList {
    public class ExpandingToSpecificNode : TreeListOperation {
        List<int> _idsToExpand;
        int currentNodeIndexInList;
        bool needsVisitChildren;
        public TreeListNode SelectedNode { get; set; }
        public ExpandingToSpecificNode(List<int> idsToExpand)
            : base() {
                needsVisitChildren = true;
            currentNodeIndexInList = 0;
            SelectedNode = null;
            _idsToExpand = idsToExpand;
        }
        public override void Execute(TreeListNode node) {
            if(currentNodeIndexInList > _idsToExpand.Count - 1) {
                return;
            }
            int idColumnValue = Convert.ToInt32(node.GetValue("ID"));
            if(idColumnValue == _idsToExpand[currentNodeIndexInList]) {
                node.Expanded = true;
                currentNodeIndexInList++;
                if(currentNodeIndexInList >= _idsToExpand.Count) {
                    SelectedNode = node;
                    needsVisitChildren = false;
                }
            }
        }
        public override bool NeedsVisitChildren(TreeListNode node) {
            return true;
        }
        public override bool NeedsFullIteration { get { return needsVisitChildren; } }
    }
}

File: VB/AsynchronousFindInTreeList/ExpandingToSpecificNode.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Data
Imports DevExpress.XtraTreeList.Nodes
Imports DevExpress.XtraTreeList
Imports System.ComponentModel
Imports System.Threading
Imports System.Windows.Forms
Imports DevExpress.XtraTreeList.Nodes.Operations

Namespace AsynchronousFindInTreeList
	Public Class ExpandingToSpecificNode
		Inherits TreeListOperation
		Private _idsToExpand As List(Of Integer)
		Private currentNodeIndexInList As Integer
		Private needsVisitChildren_Renamed As Boolean
		Private privateSelectedNode As TreeListNode
		Public Property SelectedNode() As TreeListNode
			Get
				Return privateSelectedNode
			End Get
			Set(ByVal value As TreeListNode)
				privateSelectedNode = value
			End Set
		End Property
		Public Sub New(ByVal idsToExpand As List(Of Integer))
			MyBase.New()
				needsVisitChildren_Renamed = True
			currentNodeIndexInList = 0
			SelectedNode = Nothing
			_idsToExpand = idsToExpand
		End Sub
		Public Overrides Sub Execute(ByVal node As TreeListNode)
			If currentNodeIndexInList > _idsToExpand.Count - 1 Then
				Return
			End If
			Dim idColumnValue As Integer = Convert.ToInt32(node.GetValue("ID"))
			If idColumnValue = _idsToExpand(currentNodeIndexInList) Then
				node.Expanded = True
				currentNodeIndexInList += 1
				If currentNodeIndexInList >= _idsToExpand.Count Then
					SelectedNode = node
					needsVisitChildren_Renamed = False
				End If
			End If
		End Sub
		Public Overrides Function NeedsVisitChildren(ByVal node As TreeListNode) As Boolean
			Return True
		End Function
		Public Overrides ReadOnly Property NeedsFullIteration() As Boolean
			Get
				Return needsVisitChildren_Renamed
			End Get
		End Property
	End Class
End Namespace
